core/vm/fvm/fvm.cpp

```diff
@@ -0,0 +1,263 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "vm/fvm/fvm.hpp"
+
+#include <filecoin-ffi/filcrypto.h>
+#include <shared_mutex>
+
+#include "cbor_blake/ipld_any.hpp"
+#include "cbor_blake/ipld_version.hpp"
+#include "codec/cbor/light_reader/cid.hpp"
+#include "common/ffi.hpp"
+#include "common/logger.hpp"
+#include "vm/fvm/ubig128.hpp"
+#include "vm/runtime/circulating.hpp"
+#include "vm/runtime/consensus_fault.hpp"
+#include "vm/runtime/runtime_randomness.hpp"
+#include "vm/state/impl/state_tree_impl.hpp"
+
+#define FFI_TRY(name)                                                          \
+  do {                                                                         \
+    if (res->status_code != FCPResponseStatus::FCPResponseStatus_FCPNoError) { \
+      spdlog::error(                                                           \
+          "{} status={} message={}", #name, res->status_code, res->error_msg); \
+      return ERROR_TEXT(#name);                                                \
+    }                                                                          \
+  } while (0)
+
+#define FVM_MACHINE()                               \
+  std::shared_lock lock{mutex};                     \
+  const auto machine_it{machines.find(machine_id)}; \
+  if (machine_it == machines.end()) {               \
+    return FvmError::kInvalidHandle;                \
+  }                                                 \
+  const auto machine{machine_it->second.lock()};    \
+  if (!machine) {                                   \
+    return FvmError::kInvalidHandle;                \
+  }
+
+namespace fc::vm::fvm {
+  using codec::cbor::light_reader::readCborBlake;
+
+  inline void initCallbacks();
+
+  std::shared_mutex mutex;
```

Comment:
Globally accessible variable named `fc::vm::fvm::mutex` may result in error in case of name misuse.